### PR TITLE
[Reviewer: Seb] HttpClient rework tests

### PR DESCRIPTION
**Pull request by johadalin**
In sorting out the tests for the changes in https://github.com/Metaswitch/cpp-common/pull/743, i found that the HttpClient/HttpConnection tests were a bit of a mess too, with most of the Client functionality tested through from the HttpConnection tests.
As such, this PR includes:

* A set of HttpRequest tests
* Updated ChronosConnection tests
* A re-work of the old HttpConnection tests into the HttpClient tests, now actually testing the functions, and with comments!
* A new set of HttpConnection tests. This doesn't change coverage, as it's just a header file now, but should catch any breaking changes.

We are passing full_test, and have full coverage, as before, so should be all good 👍

### CODE REVIEW

File: src/ut/chronosconnection_test.cpp [hunk 1]
```diff
@@ -15,6 +15,7 @@
 #include "utils.h"
 #include "sas.h"
 #include "fakehttpresolver.hpp"
+//#include "httpconnection.h"
```
Comment (sebrexmetaswitch) [inline src/ut/chronosconnection_test.cpp]: This can probably be deleted :)

File: src/ut/http_request_test.cpp [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+{
+  EXPECT_CALL(*_client, send_request(HttpClient::RequestType::POST, "http://server/testpath", _, _, _, _, _, _, _));
+
+  HttpRequest req = HttpRequest("server", "http", _client, HttpClient::RequestType::POST, "/testpath");
+  req.send();
+}
+
+
+// Check that the HttpRequest defaults non-mandatory arguments to empty/sensible values
+TEST_F(HttpRequestTest, test_default_params)
```
Comment (sebrexmetaswitch) [inline src/ut/http_request_test.cpp]: It's not obvious to me that yo uneed both this and the above, as they seem to both be testing that `send` is called with `POST` and `http://server/testpath`

Comment (sebrexmetaswitch) [reply src/ut/http_request_test.cpp]: We could probably remove the above, or wildcard out the method and URL from the below

File: src/ut/http_request_test.cpp [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+    delete _client; _client = NULL;
+  }
+
+  MockHttpClient* _client;
+};
+
+// Build and send a basic http request, with the default parameters
+// The send method is passed through, so we shouldn't need individual tests to
+// cover all of POST, PUT, GET, DELETE.
+TEST_F(HttpRequestTest, send_basic_request)
```
Comment (sebrexmetaswitch) [inline src/ut/http_request_test.cpp]: In general, most of our tests are CamelCase, so this would be `SendBasicRequest`. Same applies to all the below

File: src/ut/http_request_test.cpp [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+///
+// Test GET methods on HttpResponse
+///
+
+// Test HttpResponses have the return code stored properly
+TEST_F(HttpRequestTest, get_return_code)
+{
+  HTTPCode rc = HTTP_OK;
+
+  EXPECT_CALL(*_client, send_request(HttpClient::RequestType::POST, "http://server/testpath", _, _, _, _, _, _, _)).WillOnce(Return(rc));
```
Comment (sebrexmetaswitch) [inline src/ut/http_request_test.cpp]: long line

File: src/ut/http_request_test.cpp [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+  EXPECT_EQ(HTTP_OK, resp.get_return_code());
+}
+
+// Test HttpResponses have the returned body stored properly
+TEST_F(HttpRequestTest, get_resp_body)
+{
+  HTTPCode rc = HTTP_OK;
+  std::string test_resp_body = "Test body";
+
+  EXPECT_CALL(*_client, send_request(HttpClient::RequestType::POST, "http://server/testpath", _, _, _, _, _, _, _)).WillOnce(DoAll(SetArgReferee<3>(test_resp_body), Return(rc)));
```
Comment (sebrexmetaswitch) [inline src/ut/http_request_test.cpp]: long line

File: src/ut/http_request_test.cpp [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+
+// Test HttpResponses have the returned headers stored properly
+TEST_F(HttpRequestTest, get_resp_headers)
+{
+  HTTPCode rc = HTTP_OK;
+  std::map<std::string, std::string> test_resp_headers;
+  test_resp_headers.insert(std::make_pair("Test-Header", "Test value"));
+  
+
+  EXPECT_CALL(*_client, send_request(HttpClient::RequestType::POST, "http://server/testpath", _, _, _, _, _, _, _)).WillOnce(DoAll(SetArgPointee<7>(test_resp_headers), Return(rc)));
```
Comment (sebrexmetaswitch) [inline src/ut/http_request_test.cpp]: long line

File: src/ut/httpclient_test.cpp [hunk 7]
```diff
@@ -216,3 +778,232 @@ TEST_F(HttpClientTest, SasDisplayName)
 
   mock_sas_collect_messages(false);
 }
+
+// Test creation and destruction of the basic http resolver
+TEST_F(HttpClientTest, BasicResolverTest)
```
Comment (sebrexmetaswitch) [inline src/ut/httpclient_test.cpp:783]: Why is this test here? Seems completely unrelated

**Review by sebrexmetaswitch — COMMENTED** [review]
Smoe comments inline.

I like the new `HttpRequestTest`s, generally it feels like each test is testing something unique, which is great.

I've not reviewed the contents of the `HttpClientBlacklistTest`s in particular detail, as it looks like those are mostly just moved. Is that correct? If not, let me know and I'll take a more detailed look.

Also, I've not reviewed the contents of the `HttpClientTest`s in a lot of detail either. I gave these more of a look, and they seemed sensible

File: src/ut/httpclient_test.cpp (hunk 7 above)
Comment (johadalin) [reply src/ut/httpclient_test.cpp:783]: Ported across from the old httpconnection_test file, where it was if anything even more un-related. Agree, but don't know if i should really split out a whole new thing just for this.
